src/nbl/video/IPhysicalDevice.cpp

```diff
@@ -90,4 +90,375 @@ bool IPhysicalDevice::validateLogicalDeviceCreation(const ILogicalDevice::SCreat
+// If there are multiple matches: Pick smallest texel block
+// srcFormat can't be in validFormats (no promotion should be made if the format itself is valid)
+asset::E_FORMAT narrowDownFormatPromotion(const core::unordered_set<asset::E_FORMAT>& validFormats, asset::E_FORMAT srcFormat)
+{
+    asset::E_FORMAT smallestTexelBlock = asset::E_FORMAT::EF_UNKNOWN;
+    uint32_t smallestTexelBlockSize = -1;
+
+    auto srcChannels = asset::getFormatChannelCount(srcFormat);
+    assert(srcChannels);
+    auto srcAspects = getImageAspects(srcFormat);
+    bool srcIntFormat = asset::isIntegerFormat(srcFormat);
+
+    float srcPrecision[4];
+    double srcMinVal[4];
+    double srcMaxVal[4];
+    for (uint32_t channel = 0; channel < srcChannels; channel++)
+    {
+        srcPrecision[channel] = getMinFormatPrecision(srcFormat, channel);
+        srcMinVal[channel] = asset::getFormatMinValue<double>(srcFormat, channel);
+        srcMaxVal[channel] = asset::getFormatMaxValue<double>(srcFormat, channel);
+    }
+
+    // Better way to iterate the bitflags here?
+    for (auto iter = validFormats.begin(); iter != validFormats.end(); iter++)
+    {
+        asset::E_FORMAT f = *iter;
+
+        // Can't transcode to BC or planar
+        if (asset::isBlockCompressionFormat(f))
+            continue;
+        if (asset::isPlanarFormat(f))
+            continue;
+    
+        // Can't promote between int and normalized/float/scaled formats
+        if (srcIntFormat != asset::isIntegerFormat(f))
+            continue;
+
+        // Can't have less channels
+        auto dstChannels = asset::getFormatChannelCount(f);
+        if (dstChannels < srcChannels)
+            continue;
+
+        {
+            auto c = 0;
+            while (c < srcChannels)
+            {
+                // Can't have less precision (higher precision value from getFormatPrecision) than source
+                if (asset::getFormatPrecision(f, c, 0.f) > srcPrecision[c])
+                    break;
+                // Can't have less range than source
+                if (asset::getFormatMinValue<double>(f, c) > srcMinVal[c] || asset::getFormatMaxValue<double>(f, c) < srcMaxVal[c])
+                    break;
+
+                c++;
+            }
+            // If we hit a break in the while loop
+            if (c != srcChannels)
+                continue;
+        }
+
+        // Can't have less aspects
+        if (!getImageAspects(f).hasFlags(srcAspects))
+            continue;
+
+        uint32_t texelBlockSize = asset::getTexelOrBlockBytesize(f);
+        // Don't promote if we have a better valid format already
+        if (texelBlockSize > smallestTexelBlockSize) {
+            continue;
+        }
+
+        if (texelBlockSize == smallestTexelBlockSize)
+        {
+            // Tie-breaking rules:
+            // - Precision
+            // - Value range
+            if (!higherPrecisionOrValueRange(smallestTexelBlock, f, srcChannels))
```

Comment:
prefer current found fit if:
- current fit RGBA vs. BGRA matches the input format
- current fit has higher precision (but only in the source channels that are used)
- current fit has higher range (but only in the source channels that are used)

in this exact order